File: 킬링캠프/2주 차 - 리스트, 큐, 스택, 그래프, DFS, BFS/challenge/Longest Valid Parentheses/Solution.java
import java.util.*;

class Solution {
    public int longestValidParentheses(String s) {
        char[] sArr = s.toCharArray();
        Deque<Element> dq = new ArrayDeque<>();
        int result = 0;
        for (int i = 0; i < sArr.length; i++) {
            char currentValue = sArr[i];
            if (!dq.isEmpty() && (dq.peek().value == '(' && currentValue == ')')) {
                dq.pop();
                if (dq.isEmpty()) {
                    result = Math.max(result, i + 1);
                } else {
                    result = Math.max(result, i - dq.peek().index);
                }
                
                continue;
            }

            dq.push(new Element(i, currentValue));
        }

        return result;
    }
}

class Element {
    int index;
    char value;

    Element(int index, char value) {
        this.index = index;
        this.value = value;
    }
}

File: 킬링캠프/2주 차 - 리스트, 큐, 스택, 그래프, DFS, BFS/challenge/Longest Valid Parentheses/solution.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        answer = 0
        stack = [-1]

        for i, p in enumerate(s):
            if p == '(':
                stack.append(i)
                continue
                
            stack.pop()
            if not stack:
                stack.append(i)
            else:
                answer = max(answer, i - stack[-1])
        
        return answer
